Add `update_definition()` and `add_definition()` functions (#127 + #178)

elipousson opened the pull request:
## Checklist 

- [x] update NEWS.md
- [x] documentation updated with `devtools::document()`
- [ ] `devtools::check()` passes locally

## Changes 

This is a preliminary draft to add support for the updateDefinition end point #127 

**Issues that this closes** 

#127 

## Follow up tasks

- [ ] Add tests for `update_definition()` and `add_definition()`
- [ ] Add supporting functions to handle [primary](https://developers.arcgis.com/rest/services-reference/online/add-to-definition-feature-service/#primary-table) and [related tables](https://developers.arcgis.com/rest/services-reference/online/add-to-definition-feature-service/#related-tables)
- [ ] Add supporting functions to handle [adding fields](https://developers.arcgis.com/rest/services-reference/online/add-to-definition-feature-layer/#example-1) and [updating fields](https://developers.arcgis.com/rest/services-reference/online/update-definition-feature-layer/#example-eleven)

JosiahParry commented:
Can we change this to `update_definition()` to match the endpoint? The irony is that I manually crafted my first updateDefinition request yesterday lol! The tough part I think for me, is figuring out how to insert `/admin` into the URL

JosiahParry commented:
@elipousson is this ready for review? I’m going to start prepping for release in the next few days.

elipousson commented:
@JosiahParry It still needs some work before review! Not sure I'm going to get this in shape for review in advance of your clean-up for CRAN submission but I'll tag you and mark it as ready for review as soon as I have it working.

JosiahParry commented:
No worries! I just wanted to check

JosiahParry commented:
Related https://github.com/R-ArcGIS/arcgislayers/issues/178

JosiahParry commented on R/update_definition.R:
```diff
@@ -0,0 +1,117 @@
+#' @export
+update_definition <- function(
+    x,
+    update = NULL,
```
I think this is a really good start! I'm thinking we should probably use `...` here personally. That way we can `rlang::list2()` and then `lapply()` over the list to cast to json for the body

elipousson commented:
Hey @JosiahParry!

Thanks for the feedback. I dropped those helper functions so it may take users a little time to figure out the syntax but it definitely works. I figured out that I needed to use `httr2::req_body_form` to build the request and convert the `updateDefinition` object to JSON separately.

I also added some new code to display the existing definition values and updated definition values to make it easier for users to catch a mistake or typo. It may be helpful to refresh the input object by calling `arc_open` again at the end so it is returned with the updated definition values in place (rather than the old values which is how it is set up now).

JosiahParry commented on R/arc-definition.R:
```diff
@@ -0,0 +1,185 @@
+  check_resp_body_error(resp = resp)
+
+  # Refresh x to include added definitions
+  x <- arc_open(x[["url"]], token = token)
```
note that this will not work for async requests

JosiahParry commented on R/arc-definition.R:
```diff
@@ -0,0 +1,185 @@
+    path = "updateDefinition"
+  )
+
+  update_definition <- rlang::list2(...)
```
i think it would be a good idea to `arcgisutils::check_dots_named()` here to ensure that the input is named

JosiahParry replied:
```suggestion
  update_definition <- rlang::list2(...)
  arcgisutils::check_dots_names(update_definition)
```

JosiahParry commented on R/arc-definition.R:
```diff
@@ -0,0 +1,185 @@
+  x,
+  ...,
+  async = FALSE,
+  token = arc_token()
```
```suggestion
  token = arc_token()
  verbose = TRUE
```

JosiahParry commented on R/arc-definition.R:
```diff
@@ -0,0 +1,185 @@
+  definition <- x[names(x) %in% names(update_definition)]
+
+  # Display existing service/layer definition values
+  if (length(definition) > 0 && any(definition != "")) {
```
```suggestion
  if (length(definition) > 0 && any(definition != "") && verbose) {
```

JosiahParry commented on R/arc-definition.R:
```diff
@@ -0,0 +1,185 @@
+  check_resp_body_error(resp = resp)
+
+  # Display update
+  cli::cli_bullets(c("v" = "Updated definition values:"))
+  cli::cli_dl(items = update_definition)
+  cli::cli_end(dl_theme)
```
```suggestion
  if (verbose) {
    cli::cli_bullets(c("v" = "Updated definition values:"))
    cli::cli_dl(items = update_definition)
    cli::cli_end(dl_theme)
   }
```

JosiahParry commented:
This is looking really good! I've made a few changes—I've added a verbose argument and added checking to the update params.

Re the form body—I started making S7 classes and validators for this very reason (https://github.com/R-ArcGIS/arcgisutils/blob/d16fd135adf8c3d3f7fb27c03a1a0a36172ab16b/R/geoprocessing-class.R#L61) but I haven't finished—seems to be a constant for me lately lol.

@elipousson do you mind pulling this and making sure that my changes still work for your use case? If so, lets merge it.

elipousson commented:
I’ll test it and get back to you shortly.

Do we use a verbose argument anywhere else in the package? If not, how do you feel about a package-level option to suppress all messages instead? I could add a secondary message to remind about the availability of the option to disable the existing and updated definition value display that gets shown only once per session. The display could also use some work - I think it should likely show the definition terms only once and then show existing and updated side by side.

elipousson commented:
Most of the suggestions looked good, @JosiahParry.

I pulled out the `verbose` argument and moved the cli code for displaying input values into a separate helper function called `print_definition_values()`. I also got the output to look more like this: "definition name": "existing" -> "updated". 

But, by the time I finished writing up something that could handle nested list input arguments, I wasn't certain it would really be as helpful as intended. If it is easier to pull the `print_definition_values()` this PR and re-visit in the future, I'm OK with it.

In testing, I wasn't able to get `add_definition()` working but I'm not sure if it is an problem with the function or my modifications to the example JSON I pulled from the REST API documentation.

The only other feature that we could consider adding is the option to pass raw JSON. Maybe by adding `addToDefinition`, `updateDefinition`, and `deleteFromDefinition` arguments with a default value of `NULL`? Provide any of those and other parameters are ignored?

I did add a test for `update_definition()` set up like the other tests that require interactive execution.

JosiahParry commented:
I added an example that uses `add_definition()` I dont think the print method works to well here: perhaps we can add `jsonify:::pretty_json()` to it for printing? I'm going to make a follow up issue because both `add_definition()` and `update_definition()` work very well.

We can also use these as the backbone of a helper function to add / update / delete fields as well 

Here is what the rendering looks like: 

<img width="741" alt="image" src="https://github.com/user-attachments/assets/626f134f-c96c-44dd-baa1-e1bbf9349b36" />

JosiahParry commented:
@elipousson as a heads up, we have utility functions in `arcgisutils` to create layer definitions and layers—we may want to explore how to combine these two and make better documentation / vignette

``` r
library(arcgisutils)
#> 
#> Attaching package: 'arcgisutils'
#> The following object is masked from 'package:base':
#> 
#>     %||%
ld <- as_layer_definition(iris, "iris", "objectID")
l <- as_layer(iris, "iris name", "Iris Title")
str(l, 2)
#> List of 4
#>  $ featureSet     :List of 2
#>   ..$ spatialReference: Named list()
#>   ..$ features        :List of 150
#>  $ layerDefinition:List of 7
#>   ..$ name          : chr "iris name"
#>   ..$ objectIdField : chr "object_id"
#>   ..$ fields        :'data.frame':   6 obs. of  6 variables:
#>   ..$ hasAttachments: logi FALSE
#>   ..$ maxScale      : num 0
#>   ..$ minScale      : num 0
#>   ..$ type          : chr "Table"
#>  $ name           : chr "iris name"
#>  $ title          : chr "Iris Title"
str(ld, 2)
#> List of 7
#>  $ name          : chr "iris"
#>  $ objectIdField : chr "objectID"
#>  $ fields        :'data.frame':  6 obs. of  6 variables:
#>   ..$ name    : chr [1:6] "objectID" "Sepal.Length" "Sepal.Width" "Petal.Length" ...
#>   ..$ type    : chr [1:6] "esriFieldTypeOID" "esriFieldTypeDouble" "esriFieldTypeDouble" "esriFieldTypeDouble" ...
#>   ..$ alias   : chr [1:6] "objectID" "Sepal.Length" "Sepal.Width" "Petal.Length" ...
#>   ..$ length  : int [1:6] NA NA NA NA NA 255
#>   ..$ editable: logi [1:6] FALSE TRUE TRUE TRUE TRUE TRUE
#>   ..$ nullable: logi [1:6] FALSE TRUE TRUE TRUE TRUE TRUE
#>  $ hasAttachments: logi FALSE
#>  $ maxScale      : num 0
#>  $ minScale      : num 0
#>  $ type          : chr "Table"
```

<sup>Created on 2025-05-27 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>
